http.1
.\" Copyright (c) 1985, 1989, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)ftp.1	8.3 (Berkeley) 10/9/94
.\"
.\" Copyright (c) 2015 Sunil Nimmagadda <sunil@openbsd.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: August 13 2015 $
.Dt HTTP 1
.Os
.Sh NAME
.Nm http
.Nd HTTP file transfer program
.Sh SYNOPSIS
.Nm
.Op Fl ECMV
.Op Fl D Ar title
.Op Fl o Ar output
.Op Fl S Ar tls_options
.Op Fl U Ar useragent
.Op Fl w Ar seconds
.Sm off
.Pf http Oo s Oc ://
.Ar host Op : Ar port
.No / Ar file Op /
.Sm on
.Ar ...
.Sh DESCRIPTION
.Nm
allows a user to transfer files from a remote HTTP, HTTPS or an FTP server.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl C
Continue a previously interrupted file transfer.
.Nm
will continue transferring from an offset equal to the length of file.
.Pp
Resuming HTTP(S) transfers are only supported if the remote server supports the
.Dq Range
header.
.It Fl D Ar title
Specify a short title for the start of the progress bar.
.It Fl E
Disables EPSV command on IPv4 connections.
.It Fl M
Causes
.Nm
to never display the progress meter in cases where it would do so by default.
.It Fl o Ar output
When fetching a file or URL, save the contents in
.Ar output .
To make the contents go to stdout, use `-' for
.Ar output .
.It Fl S Ar tls_options
TLS options to use with HTTPS transfers.
The following settings are available:
.Bl -tag -width Ds
.It Cm cafile Ns = Ns Ar /path/to/cert.pem
PEM encoded file containing CA certificates used for certificate validation.
.It Cm capath Ns = Ns Ar /path/to/certs/
Directory containing PEM encoded CA certificates used for certificate
validation.
.It Cm ciphers Ns = Ns Ar cipher_list
Specify the list of ciphers that will be used by
.Nm .
See the
.Xr openssl 1
.Cm ciphers
subcommand.
.It Cm depth Ns = Ns Ar max_depth
Maximum depth of the certificate chain allowed when performing validation.
.It Cm dont
Don't perform server certificate validation.
.It Cm protocols Ns = Ns Ar string
Specify the TLS protocols to use.
If not specified the value
.Qq all
is used.
Refer to the
.Xr tls_config_parse_protocols 3
function for other valid protocol string values.
.It Cm muststaple
Require the server to present a valid OCSP stapling in the TLS handshake.
.It Cm noverifytime
Disable validation of certificate times and OCSP validation.
.El
.Pp
By default, server certificate validation is performed, and if it fails
.Nm
will abort.
If no
.Cm cafile
or
.Cm capath
setting is provided,
.Pa /etc/ssl/cert.pem
will be used.
.It Fl U Ar useragent
Set
.Ar useragent
as the User-Agent for HTTP(S) URL requests.
If not specified, the default User-Agent is
.Dq OpenBSD http .
.It Fl V
Disable verbose mode.
.It Fl w Ar seconds
Abort a slow connection after
.Ar seconds .
.El
.Pp
.Nm
fetches list of files specified on the command line.
The following formats are valid syntax for fetching a file:
.Bl -tag -width Ds
.Sm off
.It Xo
.Pf file:// Op Ar hostname
.No / Ar file
.Xc
.Sm on
.Dq Classic
ftp format.
.Sm off
.It Xo ftp://
.Ar host Op : Ar port
.No / Ar file
.Xc
.Sm on
An FTP URL, retrieved using the FTP protocol if
.Ev http_proxy
isn't defined.
Otherwise, transfer using HTTP via the proxy defined in
.Ev http_proxy .
.Sm off
.It Xo http://
.Ar host Op : Ar port
.No / Ar file
.Xc
.Sm on
An HTTP URL, retrieved using the HTTP protocol.
If
.Ev http_proxy
is defined, it is used as a URL to an HTTP proxy server.
.Sm off
.It Xo https://
.Ar host Op : Ar port
.No / Ar file
.Xc
.Sm on
An HTTPS URL, retrieved using the HTTPS protocol.
If
.Ev http_proxy
is defined, this HTTPS proxy server will be used to fetch the
file using the CONNECT method.
.El
.Sh ENVIRONMENT
.Nm
utilizes the following environment variables:
.Bl -tag -width Ds
.It Ev http_proxy
URL of HTTP proxy to use when making HTTP(S) URL requests.
.El
.Sh HISTORY
The
.Nm
command first appeard in
.Ox ?? .
.Sh AUTHORS
.An Sunil Nimmagadda Aq Mt sunil@openbsd.org


http.c
/*
 * Copyright (c) 2015 Sunil Nimmagadda <sunil@openbsd.org>
 * Copyright (c) 2012 - 2015 Reyk Floeter <reyk@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <err.h>
#include <libgen.h>
#include <limits.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <tls.h>

#include "http.h"

#define	DEFAULT_CA_FILE	"/etc/ssl/cert.pem"
#define MAX_REDIRECTS	10

/*
 * HTTP status codes based on IANA assignments (2014-06-11 version):
 * https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml
 * plus legacy (306) and non-standard (420).
 */
static struct http_status {
	int		 code;
	const char	*name;
} http_status[] = {
	{ 100,	"Continue" },
	{ 101,	"Switching Protocols" },
	{ 102,	"Processing" },
	/* 103-199 unassigned */
	{ 200,	"OK" },
	{ 201,	"Created" },
	{ 202,	"Accepted" },
	{ 203,	"Non-Authoritative Information" },
	{ 204,	"No Content" },
	{ 205,	"Reset Content" },
	{ 206,	"Partial Content" },
	{ 207,	"Multi-Status" },
	{ 208,	"Already Reported" },
	/* 209-225 unassigned */
	{ 226,	"IM Used" },
	/* 227-299 unassigned */
	{ 300,	"Multiple Choices" },
	{ 301,	"Moved Permanently" },
	{ 302,	"Found" },
	{ 303,	"See Other" },
	{ 304,	"Not Modified" },
	{ 305,	"Use Proxy" },
	{ 306,	"Switch Proxy" },
	{ 307,	"Temporary Redirect" },
	{ 308,	"Permanent Redirect" },
	/* 309-399 unassigned */
	{ 400,	"Bad Request" },
	{ 401,	"Unauthorized" },
	{ 402,	"Payment Required" },
	{ 403,	"Forbidden" },
	{ 404,	"Not Found" },
	{ 405,	"Method Not Allowed" },
	{ 406,	"Not Acceptable" },
	{ 407,	"Proxy Authentication Required" },
	{ 408,	"Request Timeout" },
	{ 409,	"Conflict" },
	{ 410,	"Gone" },
	{ 411,	"Length Required" },
	{ 412,	"Precondition Failed" },
	{ 413,	"Payload Too Large" },
	{ 414,	"URI Too Long" },
	{ 415,	"Unsupported Media Type" },
	{ 416,	"Range Not Satisfiable" },
	{ 417,	"Expectation Failed" },
	{ 418,	"I'm a teapot" },
	/* 419-421 unassigned */
	{ 420,	"Enhance Your Calm" },
	{ 422,	"Unprocessable Entity" },
	{ 423,	"Locked" },
	{ 424,	"Failed Dependency" },
	/* 425 unassigned */
	{ 426,	"Upgrade Required" },
	/* 427 unassigned */
	{ 428,	"Precondition Required" },
	{ 429,	"Too Many Requests" },
	/* 430 unassigned */
	{ 431,	"Request Header Fields Too Large" },
	/* 432-450 unassigned */
	{ 451,	"Unavailable For Legal Reasons" },
	/* 452-499 unassigned */
	{ 500,	"Internal Server Error" },
	{ 501,	"Not Implemented" },
	{ 502,	"Bad Gateway" },
	{ 503,	"Service Unavailable" },
	{ 504,	"Gateway Timeout" },
	{ 505,	"HTTP Version Not Supported" },
	{ 506,	"Variant Also Negotiates" },
	{ 507,	"Insufficient Storage" },
	{ 508,	"Loop Detected" },
	/* 509 unassigned */
	{ 510,	"Not Extended" },
	{ 511,	"Network Authentication Required" },
	/* 512-599 unassigned */
	{ 0,	NULL },
	};

struct http_headers {
	char	*location;
	off_t	 content_length;
};

static void		 headers_parse(int);
static void		 http_close(struct url *);
static const char	*http_error(int);
static struct url	*http_redirect(struct url *, char *);
static int		 http_status_code(const char *);
static int		 http_status_cmp(const void *, const void *);
static int		 http_request(int, const char *, ...)
			    __attribute__((__format__ (printf, 2, 3)))
			    __attribute__((__nonnull__ (2)));
static ssize_t		 tls_getline(char **, size_t *, struct tls *);
static char		*relative_path_resolve(const char *,
			    const char *);

static struct http_headers	 headers;
static struct tls_config	*tls_config;
static struct tls		*ctx;
static FILE			*fp;

static char * const		 tls_verify_opts[] = {
#define HTTP_TLS_CAFILE		0
	"cafile",
#define HTTP_TLS_CAPATH		1
	"capath",
#define HTTP_TLS_CIPHERS	2
	"ciphers",
#define HTTP_TLS_DONTVERIFY	3
	"dont",
#define HTTP_TLS_VERIFYDEPTH	4
	"depth",
#define HTTP_TLS_PROTOCOLS	5
	"protocols",
#define HTTP_TLS_MUSTSTAPLE	6
	"muststaple",
#define HTTP_TLS_NOVERIFYTIME	7
	"noverifytime",
	NULL
};

void
https_init(void)
{
	char		*str;
	int		 depth;
	uint32_t	 http_tls_protocols;
	const char	*ca_file = DEFAULT_CA_FILE, *errstr;

	if (tls_init() != 0)
		errx(1, "tls_init failed");

	if ((tls_config = tls_config_new()) == NULL)
		errx(1, "tls_config_new failed");

	while (tls_options && *tls_options) {
		switch (getsubopt(&tls_options, tls_verify_opts, &str)) {
		case HTTP_TLS_CAFILE:
			if (str == NULL)
				errx(1, "missing CA file");
			ca_file = str;
			break;
		case HTTP_TLS_CAPATH:
			if (str == NULL)
				errx(1, "missing ca path");
			if (tls_config_set_ca_path(tls_config, str) != 0)
				errx(1, "tls ca path failed");
			break;
		case HTTP_TLS_CIPHERS:
			if (str == NULL)
				errx(1, "missing cipher list");
			if (tls_config_set_ciphers(tls_config, str) != 0)
				errx(1, "tls set ciphers failed");
			break;
		case HTTP_TLS_DONTVERIFY:
			tls_config_insecure_noverifycert(tls_config);
			tls_config_insecure_noverifyname(tls_config);
			break;
		case HTTP_TLS_PROTOCOLS:
			if (tls_config_parse_protocols(&http_tls_protocols,
			    str) != 0)
				errx(1, "tls parsing protocols failed");
			tls_config_set_protocols(tls_config,
			    http_tls_protocols);
			break;
		case HTTP_TLS_VERIFYDEPTH:
			if (str == NULL)
				errx(1, "missing depth");
			depth = strtonum(str, 0, INT_MAX, &errstr);
			if (errstr)
				errx(1, "Cert validation depth is %s", errstr);
			tls_config_set_verify_depth(tls_config, depth);
			break;
		case HTTP_TLS_MUSTSTAPLE:
			tls_config_ocsp_require_stapling(tls_config);
			break;
		case HTTP_TLS_NOVERIFYTIME:
			tls_config_insecure_noverifytime(tls_config);
			break;
		default:
			errx(1, "Unknown -S suboption `%s'",
			    suboptarg ? suboptarg : "");
		}
	}

	if (tls_config_set_ca_file(tls_config, ca_file) == -1)
		errx(1, "tls_config_set_ca_file failed");
}

void
http_connect(struct url *url, int timeout)
{
	int	sock;

	sock = tcp_connect(url->host, url->port, timeout);
	if ((fp = fdopen(sock, "r+")) == NULL)
		err(1, "%s: fdopen", __func__);

	if (proxy)
		proxy_connect(url, fp);

	if (url->scheme == S_HTTP)
		return;

	if ((ctx = tls_client()) == NULL)
		errx(1, "failed to create tls client");

	if (tls_configure(ctx, tls_config) != 0)
		errx(1, "%s: %s", __func__, tls_error(ctx));

	if (tls_connect_socket(ctx, sock, url->host) != 0)
		errx(1, "%s: %s", __func__, tls_error(ctx));
}

void
proxy_connect(struct url *url, FILE *proxy_fp)
{
	int	code;

	code = http_request(url->scheme,
	    "CONNECT %s:%s HTTP/1.0\r\n"
	    "Host: %s\r\n"
	    "User-Agent: %s\r\n"
	    "\r\n",
	    url->host,
	    url->port,
	    url->host,
	    ua);

	if (code != 200)
		errx(1, "%s: failed to CONNECT to %s:%s: %s",
		    __func__, url->host, url->port, http_error(code));
}

struct url *
http_get(struct url *url)
{
	char			*range;
	int			 code, redirects = 0;

 redirected:
	if (asprintf(&range, "Range: bytes=%lld-\r\n", url->offset) == -1)
		err(1, "%s: asprintf", __func__);

	code = http_request(url->scheme,
	    "GET %s HTTP/1.0\r\n"
	    "Host: %s\r\n"
	    "User-Agent: %s\r\n"
	    "%s"
	    "\r\n",
	    url->path ? url->path : "/",
	    url->host,
	    ua,
	    url->offset ? range : "");
	free(range);

	switch (code) {
	case 200:
		if (url->offset)
			warnx("Server does not support resume.");
		url->offset = 0;
		break;
	case 206:
		break;
	case 301:
	case 302:
	case 303:
	case 307:
		http_close(url);
		if (++redirects > MAX_REDIRECTS)
			errx(1, "Too many redirections requested");

		if (headers.location == NULL)
			errx(1, "%s: Location header missing", __func__);

		url = http_redirect(url, headers.location);
		log_request("Redirected to", url);
		http_connect(url, 0);
		log_request("Requesting", url);
		goto redirected;
	case 416:
		warnx("File is already fully retrieved");
		break;
	default:
		errx(1, "Error retrieving file: %d %s", code, http_error(code));
	}

	url->file_sz = headers.content_length + url->offset;
	return url;
}

static struct url *
http_redirect(struct url *old_url, char *location)
{
	struct url	*new_url;
	const char	*http, *https;

	http = scheme_str[S_HTTP];
	https = scheme_str[S_HTTPS];

	if (strncasecmp(location, http, strlen(http)) == 0 ||
	    strncasecmp(location, https, strlen(https)) == 0) {
		/* absolute uri reference */
		new_url = url_parse(location);
		if (old_url->scheme == S_HTTPS && new_url->scheme != S_HTTPS)
			errx(1, "aborting HTTPS to HTTP redirect");
	} else {
		/* relative uri reference */
		if ((new_url = calloc(1, sizeof *new_url)) == NULL)
			err(1, "%s: calloc", __func__);

		new_url->scheme = old_url->scheme;
		new_url->host = xstrdup(old_url->host, __func__);
		new_url->port = xstrdup(old_url->port, __func__);

		 /* absolute-path reference */
		if (location[0] == '/')
			new_url->path = xstrdup(location, __func__);
		else {
			new_url->path = relative_path_resolve(old_url->path,
			    location);
		}
	}

	new_url->fname = xstrdup(old_url->fname, __func__);
	url_free(old_url);
	return new_url;
}

static char *
relative_path_resolve(const char *base_path, const char *location)
{
	char	*new_path, *p;

	/* trim fragment component from both uri */
	if ((p = strchr(location, '#')) != NULL)
		*p = '\0';
	if (base_path && (p = strchr(base_path, '#')) != NULL)
		*p = '\0';

	if (base_path == NULL) {
		if (asprintf(&new_path, "/%s", base_path) == -1)
			err(1, "%s: asprintf", __func__);
	} else if (base_path[strlen(base_path) - 1] == '/') {
		if (asprintf(&new_path, "%s%s", base_path, location) == -1)
			err(1, "%s: asprintf", __func__);
	} else {
		p = dirname(base_path);
		if (asprintf(&new_path, "%s/%s",
		    strcmp(p, ".") == 0 ? "" : p, location) == -1)
			err(1, "%s: asprintf", __func__);
	}

	return new_path;
}

void
http_save(struct url *url, int fd)
{
	FILE	*dst_fp;
	char	*tmp_buf;
	ssize_t	 r;

	if ((dst_fp = fdopen(fd, "w")) == NULL)
		err(1, "%s: fdopen", __func__);

	if (url->scheme == S_HTTP) {
		copy_file(url, fp, dst_fp);
		goto done;
	}

	if ((tmp_buf = malloc(TMPBUF_LEN)) == NULL)
		err(1, "%s: malloc", __func__);

	for (;;) {
		do {
			r = tls_read(ctx, tmp_buf, TMPBUF_LEN);
		} while (r == TLS_WANT_POLLIN || r == TLS_WANT_POLLOUT);

		if (r == -1)
			err(1, "%s: tls_read: %s", __func__, tls_error(ctx));
		else if (r == 0)
			break;

		url->offset += r;
		if (fwrite(tmp_buf, 1, r, dst_fp) != r)
			err(1, "%s: fwrite", __func__);
	}
	free(tmp_buf);

 done:
 	fclose(dst_fp);
	http_close(url);
}

static void
http_close(struct url *url)
{
	ssize_t	r;

	if (url->scheme == S_HTTPS) {
		do {
			r = tls_close(ctx);
		} while (r == TLS_WANT_POLLIN || r == TLS_WANT_POLLOUT);
		tls_free(ctx);
	}

	fclose(fp);
}

static int
http_request(int scheme, const char *fmt, ...)
{
	va_list	 ap;
	char	*req, *buf = NULL;
	size_t	 n = 0;
	ssize_t	 nw;
	int	 code, r;

	va_start(ap, fmt);
	r = vasprintf(&req, fmt, ap);
	va_end(ap);
	if (r < 0)
		err(1, "%s: vasprintf", __func__);

	if (http_debug)
		fprintf(stderr, "<<< %s\n", req);

	if (scheme == S_HTTP) {
		if (fprintf(fp, "%s\r\n", req) < 0)
			errx(1, "%s: fprintf", __func__);
		(void)fflush(fp);
		if (getline(&buf, &n, fp) == -1)
			err(1, "%s: getline", __func__);
	} else {
		do {
			nw = tls_write(ctx, req, r);
		} while (nw == TLS_WANT_POLLIN || nw == TLS_WANT_POLLOUT);
		if (nw == -1)
			errx(1, "%s: tls_write", __func__);
		if (tls_getline(&buf, &n, ctx) == -1)
			errx(1, "%s: tls_getline", __func__);
	}

	free(req);
	if (http_debug)
		fprintf(stderr, ">>> %s", buf);

	if ((code = http_status_code(buf)) == -1)
		errx(1, "%s: failed to extract status code", __func__);

	free(buf);
	headers_parse(scheme);
	return code;
}

static int
http_status_code(const char *status_line)
{
	unsigned int	code;

	if (sscanf(status_line, "%*s %u %*s", &code) != 1)
		errx(1, "%s: failed to extract status code", __func__);

	if (code < 100 || code > 511)
		errx(1, "%s: invalid status code %d", __func__, code);

	return code;
}

static void
headers_parse(int scheme)
{
	char			*buf = NULL, *p;
	const char		*e;
	size_t			 n = 0;
	ssize_t			 buflen;

	free(headers.location);
	memset(&headers, 0, sizeof(headers));
	for (;;) {
		if (scheme == S_HTTP) {
			if ((buflen = getline(&buf, &n, fp)) == -1)
				err(1, "%s: getline", __func__);
		} else {
			if ((buflen = tls_getline(&buf, &n, ctx)) == -1)
				errx(1, "%s: tls_getline", __func__);
		}

		buf[buflen - 1] = '\0';
		buflen -= 1;
		if (buflen > 0 && buf[buflen - 1] == '\r') {
			buf[buflen - 1] = '\0';
			buflen -= 1;
		}

		if (http_debug)
			fprintf(stderr, ">>> %s\n", buf);

		if (buflen == 0)
			break; /* end of headers */

		if (strncasecmp(buf, "Content-Length: ", 16) == 0) {
			if ((p = strchr(buf, ' ')) == NULL)
				errx(1, "Failed to parse Content-Length");

			p++;
			headers.content_length = strtonum(p, 0, INT64_MAX, &e);
			if (e)
				err(1, "%s: Content Length is %s: %lld",
				    __func__, e, headers.content_length);
		}

		if (strncasecmp(buf, "Location: ", 10) == 0) {
			if ((p = strchr(buf, ' ')) == NULL)
				errx(1, "Failed to parse Location");

			headers.location = xstrdup(++p, __func__);
		}

	}

	free(buf);
}

static const char *
http_error(int code)
{
	struct http_status	error, *res;

	/* Set up key */
	error.code = code;

	if ((res = bsearch(&error, http_status,
	    sizeof(http_status) / sizeof(http_status[0]) - 1,
	    sizeof(http_status[0]), http_status_cmp)) != NULL)
		return (res->name);

	return (NULL);
}

static int
http_status_cmp(const void *a, const void *b)
{
	const struct http_status *ea = a;
	const struct http_status *eb = b;

	return (ea->code - eb->code);
}

#define MINBUF	128
static ssize_t
tls_getline(char **buf, size_t *buflen, struct tls *tls)
{
	char		*newb;
	size_t		 newlen, off;
	int		 ret;
	unsigned char	 c;

	if (buf == NULL || buflen == NULL) {
		/* tls_set_errorx(tls, "invalid arguments"); */
		return -1;
	}

	/* If buf is NULL, we have to assume a size of zero */
	if (*buf == NULL)
		*buflen = 0;

	off = 0;
	do {
		do {
			ret = tls_read(tls, &c, 1);
		} while (ret == TLS_WANT_POLLIN || ret == TLS_WANT_POLLOUT);
		if (ret == -1)
			return -1;

		/* Ensure we can handle it */
		if (off + 2 > SSIZE_MAX) {
			/* tls_set_errorx(tls, "overflow"); */
			return -1;
		}

		newlen = off + 2; /* reserve space for NUL terminator */
		if (newlen > *buflen) {
			newlen = newlen < MINBUF ? MINBUF : *buflen * 2;
			newb = recallocarray(*buf, *buflen, newlen, 1);
			if (newb == NULL) {
				/* tls_set_error(tls, "reallocarray"); */
				return -1;
			}
			*buf = newb;
			*buflen = newlen;
		}

		*(*buf + off) = c;
		off += 1;
 	} while (c != '\n');

	*(*buf + off) = '\0';
	return off;
}
